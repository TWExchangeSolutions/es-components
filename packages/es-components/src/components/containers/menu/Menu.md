Use the `Menu` to expose an list of items overlayed over the screen when a button is pressed. All nodes should be `Menu.MenuSection`s

### Basic Example
```
<Menu header="Small Menu" buttonContent="Open Menu" openButtonType="primary" rootClose hasBackdrop>
  <Menu.MenuSection title="Menu Section" isFirst rootClose>
    <a href="http://www.google.com">Go To Google</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 2">
    <a href="http://www.yahoo.com">Go To Yahoo</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 3">
    <a href="http://www.lycos.com">Go To Lycos</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 4">
    <a href="http://www.dogpile.com">Go To Dogpile</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 5">
    <a href="http://www.bing.com">Go To Bing</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 6">
    <a href="http://www.ask.com">Go To Ask</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 7">
    <a href="http://www.search.com">Go To Search</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 8" isLast>
    <a href="http://www.duckduckgo.com">Go To Duck Duck Go</a>
  </Menu.MenuSection>
</Menu>
```

### Inline Example
By adding the `Inline` prop the menu will be displayed without bottom borders side by side.

```
<Menu header="Small Menu" buttonContent="Open Menu" inline>
  <Menu.MenuSection title="Menu Section" isFirst>
    <a href="http://www.google.com">Go To Google</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 2">
    <a href="http://www.yahoo.com">Go To Yahoo</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 3">
    <a href="http://www.lycos.com">Go To Lycos</a>
  </Menu.MenuSection>
</Menu>
```