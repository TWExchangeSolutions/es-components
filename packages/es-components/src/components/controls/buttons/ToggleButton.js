/* eslint no-confusing-arrow: 0 */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import Button from './Button';
import OutlineButton from './OutlineButton';
import { useTheme } from '../../util/useTheme';

const buttonMixin = css`
  background-color: ${props =>
    props.isPressed ? props.variant.hoverBgColor : props.variant.bgColor};
  color: ${props =>
    props.isPressed ? props.variant.activeTextColor : props.variant.textColor};
`;

const StyledButton = styled(Button)`
  ${buttonMixin};
`;

const StyledOutlineButton = styled(OutlineButton)`
  ${buttonMixin};
`;

function ToggleButton(props) {
  const theme = useTheme();
  const [isPressed, setIsPressed] = useState(props.isPressed);

  function toggleButton(event) {
    setIsPressed(!isPressed);
    props.onClick(event);
  }

  const { styleType, size, block, isOutline, ...buttonProps } = props;

  if (isOutline) {
    return (
      <StyledOutlineButton
        {...buttonProps}
        onClick={toggleButton}
        styleType={styleType}
        size={size}
        block={block}
        isOutline={isOutline}
        isPressed={isPressed}
        variant={theme.buttonStyles.outlineButton.variant[styleType]}
      >
        {props.children}
      </StyledOutlineButton>
    );
  }

  return (
    <StyledButton
      {...buttonProps}
      onClick={toggleButton}
      styleType={styleType}
      size={size}
      block={block}
      isOutline={isOutline}
      isPressed={isPressed}
      variant={theme.buttonStyles.button.variant[styleType]}
    >
      {props.children}
    </StyledButton>
  );
}

ToggleButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  styleType: PropTypes.string,
  size: PropTypes.oneOf(['lg', 'default', 'sm', 'xs']),
  block: PropTypes.bool,
  isOutline: PropTypes.bool,
  isPressed: PropTypes.bool
};

ToggleButton.defaultProps = {
  styleType: 'default',
  size: 'default',
  block: false,
  isOutline: false,
  isPressed: false
};

export default ToggleButton;
