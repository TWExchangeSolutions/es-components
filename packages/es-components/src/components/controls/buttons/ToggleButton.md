A simple button that maintains the "depressed" state/display when clicked.

```
function noop() { }
<div>
  <span style={{marginRight: '15px'}}>
    <ToggleButton size="lg" styleType="primary" onClick={noop}>Toggle Me!</ToggleButton>
  </span>
  <ToggleButton styleType="success" isOutline onClick={noop}>Toggle Me!</ToggleButton>
</div>
```
