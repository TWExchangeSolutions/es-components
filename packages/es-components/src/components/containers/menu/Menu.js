import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';
import ToggleButton from '../../controls/buttons/ToggleButton';
import MenuPanel from './MenuPanel';
import MenuSection from './MenuSection';
import RootCloseWrapper from 'react-overlays/lib/RootCloseWrapper';

const Backdrop = styled.div`
  background-color: black;
  bottom: 0;
  cursor: auto;
  left: 0;
  opacity: 0.5;
  position: fixed;
  right: 0;
  top: 0;
  z-index: auto;
  display: ${props => (props.isMenuOpen ? 'inherit' : 'none')};
`;

export class Menu extends React.Component {
  static childContextTypes = {
    inline: PropTypes.bool
  };

  state = {
    isMenuOpen: false
  };

  getChildContext() {
    return { inline: this.props.inline };
  }

  toggleMenu = () => {
    this.setState(previousState => ({ isMenuOpen: !previousState.isMenuOpen }));
  };

  closeMenu = () => {
    if (this.state.isMenuOpen) {
      this.setState({ isMenuOpen: false });
    }
  };

  render() {
    const {
      children,
      buttonContent,
      className,
      openButtonType,
      rootClose,
      hasBackdrop,
      headerContent
    } = this.props;

    return (
      <RootCloseWrapper onRootClose={this.closeMenu} disabled={!rootClose}>
        <div className={classnames('es-menu', className)}>
          {hasBackdrop && (
            <Backdrop
              className="es-menu__backdrop"
              isMenuOpen={this.state.isMenuOpen}
              onClick={this.closeMenu}
            />
          )}
          <ToggleButton
            handleOnClick={this.toggleMenu}
            isPressed={this.state.isMenuOpen}
            styleType={openButtonType}
            aria-expanded={this.state.isMenuOpen}
          >
            {buttonContent}
          </ToggleButton>
          <MenuPanel
            headerContent={headerContent}
            isOpen={this.state.isMenuOpen}
            onClose={this.closeMenu}
          >
            {children}
          </MenuPanel>
        </div>
      </RootCloseWrapper>
    );
  }
}

Menu.MenuSection = MenuSection;

Menu.propTypes = {
  children: PropTypes.any.isRequired,
  buttonContent: PropTypes.any.isRequired,
  openButtonType: PropTypes.string,
  className: PropTypes.string,
  rootClose: PropTypes.bool,
  inline: PropTypes.bool,
  hasBackdrop: PropTypes.bool,
  headerContent: PropTypes.node
};

Menu.defaultProps = {
  rootClose: false,
  openButtonType: 'default'
};

export default Menu;
