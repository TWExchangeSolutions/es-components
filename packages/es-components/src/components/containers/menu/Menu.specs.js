/* eslint-env jest */

import React from 'react';
import { mountWithTheme, renderWithTheme } from '../../../testing';
import { Menu } from './Menu';
import ToggleButton from '../../controls/buttons/ToggleButton';

describe('MenuTestSuite', () => {
  let instanceToRender;

  beforeEach(() => {
    instanceToRender = (
      <Menu
        headerContent="Small Menu"
        buttonContent="Open Menu"
        className="test"
      >
        <Menu.MenuSection title="Menu Section" isFirst>
          <a href="www.google.com">Go To Google</a>
        </Menu.MenuSection>
      </Menu>
    );
  });

  it('renders as expected', () => {
    const tree = renderWithTheme(instanceToRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('toggles menu open/closed on click', () => {
    const instance = mountWithTheme(instanceToRender);
    const button = instance.find(ToggleButton);

    button.simulate('click');
    expect(instance.state().isMenuOpen).toBe(true);
    button.simulate('click');
    expect(instance.state().isMenuOpen).toBe(false);
  });
});
