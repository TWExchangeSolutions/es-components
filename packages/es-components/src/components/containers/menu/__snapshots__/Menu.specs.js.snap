// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuTestSuite renders as expected 1`] = `
<div
  className={undefined}
>
  <button
    aria-expanded={false}
    className="open-menu-button sc-ifAKCX drovqE sc-bwzfXH kqCNDZ sc-bdVaJa jXnGOW"
    onClick={[Function]}
  >
    Open Menu
  </button>
  <div
    className="menu-panel sc-gzVnrw ePSKQk"
  >
    <div
      className="sc-dnqmqq djHNQY"
    >
      <button
        aria-label="Close"
        className="sc-htoDjs hubgrr sc-EHOje cQHrKQ"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          Ã—
        </span>
        <span
          className="sc-bZQynM eBGzNP"
        >
          Close
        </span>
      </button>
    </div>
    <div
      className="sc-iwsKbI kyjaoD"
    >
      <div
        className="sc-gZMcBi ivZJCw"
      >
        <div
          aria-label="Menu Section"
          className="sc-gqjmRU fiUekS"
        >
          Menu Section
        </div>
        <div
          className="sc-VigVT frPBLq"
        >
          <a
            href="www.google.com"
          >
            Go To Google
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
